docs/gallery/plot_generic_data_chunk_tutorial.py

```diff
@@ -0,0 +1,118 @@
+my_array = np.random.randint(low=0, high=10, size=(20000, 5000), dtype="int32")
+my_custom_iterator = NumpyArrayDataChunkIterator(array=my_array, buffer_gb=0.2)
+
+for j, buffer in enumerate(my_custom_iterator, start=1):
+    print(f"Buffer number {j} returns data from selection: {buffer.selection}")
+
+###############################################################################
+# .. note::
+#   Technically, in this example the total data is still fully loaded into RAM from the initial Numpy array.
+#   A more accurate use case would be achieved from writing the test_array to a temporary file on your system
+#   and loading it back with np.memmap, which is a subtype of Numpy arrays that do not immediately load the data.
+
+###############################################################################
+# Writing to an HDF5 file with full control of shape arguments
+# ------------------------------------------------------------
+# The true intention of returning data selections of this form, and within a DataChunk object,
+# is to write these piecewise to an HDF5 dataset.
+
+# This is where the importance of the underlying `chunk_shape` comes in, and why it is critical to performance
+# that it perfectly subsets the `buffer_shape`.
+import h5py
+
+maxshape = (20000, 5000)
+buffer_shape = (10000, 2500)
+chunk_shape = (1000, 250)
+
+my_array = np.random.randint(low=0, high=10, size=maxshape, dtype="int32")
+my_custom_iterator = NumpyArrayDataChunkIterator(array=my_array, buffer_shape=buffer_shape, chunk_shape=chunk_shape)
+with h5py.File(name="my_temporary_test_file.hdf5", mode="w") as f:
+    dset = f.create_dataset(name="test", shape=maxshape, dtype="int16", chunks=my_custom_iterator.chunk_shape)
+    for buffer in my_custom_iterator:
+        dset[buffer.selection] = buffer.data
+# Remember to remove the temporary file after running this and exploring the contents!
+
+###############################################################################
+# .. note::
+#   Here we explicitly set the `chunks` value in the HDF5 dataset object; however, a nice part of the design of this
+#   iterator is that when wrapped in a ``hdmf.backends.hdf5.h5_utils.H5DataIO`` that is called within a
+#  ``hdmf.backends.hdf5.h5tools.HDF5IO`` context with a corresponding ``hdmf.container.Container``, these details
+#  will be automatically parsed.
+
+###############################################################################
+# .. note::
+#   There may be conflicts of nomenclature found between HDMF and HDF5. 'Chunks' in the HDF5 dataset are the tiny
+#   pieces of data that get compressed or cached together, and they should always be less than 1 megabyte (MB) in
+#   size for optimal performance, in constrast to the 'buffer' of the iterator whose size which should be as much
+#   free RAM as can be safely used. This buffer utilizes 'DataChunks' in the HDMF sense, which is any piece of
+#   data smaller than the full extent of the maximum shape.
```

Comment:
```suggestion
###############################################################################
# .. note::
#   There is some overlap here in nomenclature between HDMF and HDF5. The term *chunk* in both
#   HDMF and HDF5 refer to a subset of dataset, however, in HDF5 a chunk is a piece of dataset on disk,
#   whereas in the context of the data chunk iteration a chunk is a block data in memory. As such, the 
#   requirements on the shape and size of chunks are different. In HDF5  chunks are typically small pieces
#   of a dataset that get compressed and cached together, and they should usually be small in size for 
#   optimal performance  (typically 1 MB or less). In contrast, a :py:class:`~hdmf.data_utils.DataChunk` in 
#   HDMF acts as a block of data for writing data to dataset. To improve performance and  avoid repeat 
#   updates to the same `Chunk` in the HDF5 file, :py:class:`~hdmf.data_utils.DataChunk` objects for write 
#   should align with `Chunks` in the HDF5 file, i.e., the :py:class:`~hdmf.data_utils.DataChunk.selection`   
#   should fully cover one or more `Chunks`  in the HDF5 file to avoid repeat updates to the same
#   `Chunks` in the HDF5 file. The `buffer` of the iterator then collects multiple 
#   :py:class:`~hdmf.data_utils.DataChunk` objects to help reduce the number of small I/O operations 
#   and help improve performance. The `buffer` should usually be large, i.e, be able to use
#   as much free RAM as can be safely used. 
```

How does this sound?

Reply:
we have changed from "chunk" to "buffer"

Reply:
I don't think the buffer collections multiple :py:class:`~hdmf.data_utils.DataChunk` objects, I think it's just one. We are using the term "buffer" to try to make this distinction more clear.